### PR TITLE
Remove normals from primitives

### PR DESCRIPTION
**Pull request by fsimonis**
## Main changes of this PR

Removes normals from Primitives and computes them on-demand if necessary.

- Remove saved normals from primitives
- Remove vertex normals from vtk exporters
- Remove `Mesh::computeState`
- Compute normals on-demand
- Extended and fixed tests

Deprecated the `normal` argument of the `vertexCallback(id, coords, normal):`.
If there is a 3rd parameter, preCICE will display the following warning and pass `None` as a value:
```
action::PythonAction|l253|initialize|WARNING: Python module "TestDeprecatedAction" defines the function vertexCallback with 3 arguments. The normal argument is deprecated and preCICE will pass None instead. Please use the following definition to silence this warning "def vertexCallback(id, coords):"
```

Primitive | Mem Old | Mem New | Rel
--- | --- | --- | ---
Vertex | 96B | 32B + 24B | ~41%
Edge | 64B | 24B | ~62%
Triangle | 72B | 32B | ~55%

## Motivation and additional information

Normals are saved as `Eigen::VectorXd`, which are `16B` objects with `24B` or `16B` dynamically allocated data.
This memory is used for every Vertex, Edge and Triangle in preCICE.
You can find more information on the memory consumption in #483.

We can compute the Edge and Triangle normals on-demand.
The vertex normals are unused in most cases.
This PR also deprecates the vertex normals in the python actions and removes them from the exporters.

## Author's checklist

* [x] I added a changelog file with this PR number in `docs/changelog/` if there are noteworthy changes.
* [x] I ran `tools/formatting/check-format` and everything is formatted correctly.
* [x] I sticked to C++14 features.
* [x] I sticked to CMake version 3.10.
* [ ] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [x] Does the changelog entry make sense? Is it formatted correctly?
* [x] Do you understand the code changes?
* [ ] (more questions/tasks)

### CODE REVIEW

**Comment by fsimonis** [comment]
This PR removes the vertex normals from the python bindings. Do we consider this a breaking change?

**Comment by BenjaminRodenberg** [comment]
> This PR removes the vertex normals from the python ~~bindings~~ **actions**. Do we consider this a breaking change?

@fsimonis ?

**Comment by fsimonis** [comment]
I opened [a poll on our discourse](https://precice.discourse.group/t/user-poll-on-vertex-normals/574) regarding the breaking nature of this change.

File: src/mesh/tests/MeshTest.cpp [hunk 1]
```diff
@@ -29,96 +29,6 @@ BOOST_AUTO_TEST_SUITE(MeshTests)
...
-  //     *  <---
-  // *****
-  Edge &e2 = mesh.createEdge(v2, v3);
-  mesh.computeState();
-
-  // Perform test validations
-  BOOST_TEST(equals(e1.getCenter(), Vector2d(0.5, 0.0)));
-  BOOST_TEST(equals(e2.getCenter(), Vector2d(1.0, 0.5)));
-  BOOST_TEST(e1.getEnclosingRadius() == 0.5);
-  BOOST_TEST(e2.getEnclosingRadius() == 0.5);
```
Comment (uekerman) [inline src/mesh/tests/MeshTest.cpp:54]: I think we are losing all (unit) tests on these functions. Could you please add them to `EdgeTest` and `TriangleTest`.

**Review by uekerman — APPROVED** [review]
I like it. Let's do this :+1:

In particular, great that the confusing `computeState` is now gone :rocket:  

Now, that we are touching all this already: I think we can also deprecate the flipping of normals. As I see it this is nowhere used, right? Or does the NP mapping use it?

When merging please open an issue to remove the deprecated function + test + configuration with the v3.0 milestone.

Only minor things below.

**Comment by fsimonis** [comment]
According to [the poll](https://precice.discourse.group/t/user-poll-on-vertex-normals/574),  we have at least one user that relies on the normals in vtk exports.

@uekerman A possible solution is to implement this feature in the vtk exporter and only calculate the normals if it is enabled. We could even make the normals configurable in cases one needs the export, but cannot afford the extra memory.

File: src/mesh/Edge.cpp [hunk 2]
```diff
@@ -30,17 +29,19 @@ double Edge::getLength() const
   return length;
 }
 
-const Eigen::VectorXd Edge::computeNormal(bool flip)
+Eigen::VectorXd Edge::computeNormal(bool flip) const
 {
   // Compute normal
   Eigen::VectorXd edgeVector = vertex(1).getCoords() - vertex(0).getCoords();
-  Eigen::VectorXd normal     = Eigen::Vector2d(-edgeVector[1], edgeVector[0]);
+  // In 3D, use a normal on the plane z=0
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply src/mesh/Edge.cpp:36]: In 3D, the nearest projection will project onto the edge if there is no valid triangle.
This currently requires the normal of the edge in 3D. As there are infinite choices, it is easiest to simply select a normal on the z plane. I just realized that this doesn't work for an edgeVector of `(0,0,1)` though.

https://github.com/precice/precice/blob/813de05a87270ae9194bb4a499895a9fab4caef9/src/math/barycenter.cpp#L13-L17

[the remark this answers is not in the document]
Comment (uekerman) [reply src/mesh/Edge.cpp:36]: Sounds a bit strange that the NP mapping then needs just any vector in the normal plane. Wouldn't it be better to just directly pass the edge vector to the mapping (i.e. the normal plane)?

**Comment by fsimonis** [comment]
@uekerman I propose to refactor the projection logic in a future PR.

**Review by uekerman — APPROVED** [review]
Good to go for me.

> When merging please open an issue to remove the deprecated function + test + configuration with the v3.0 milestone.

Please don't forget

> @uekerman I propose to refactor the projection logic in a future PR.

OK. The current solution is a bit hacky. Ideally, calling `Edge::computeNormal` should only be allowed in 2D.

**Pull request merged**